src/storage-client/src/controller.rs

```diff
@@ -1182,18 +1182,48 @@ where
         let mut to_create = Vec::with_capacity(to_register.len());
         // This work mutates the controller state, so must be done serially. Because there
         // is no io-bound work, its very fast.
-        for (id, description, write, since_handle, metadata) in to_register {
+        for (id, description, mut write, since_handle, metadata) in to_register {
             let storage_dependencies = description.get_storage_dependencies();
 
             let data_shard_since = since_handle.since().clone();
             let dependency_since = self.determine_dependency_since(&storage_dependencies)?;
             let combined_since = data_shard_since.join(&dependency_since);
+
+            // Ensure out write handle's upper is in advance of the combined
+            // since unless they're being initialized.
+            let mut upper = write.upper().clone();
+            while PartialOrder::less_equal(&upper, &combined_since)
+                && upper.as_option() != Some(&T::minimum())
+            {
+                let since = combined_since.as_option().expect("collection not closed");
```

Comment:
I don't think we should be appending empty updates here. The best thing we can do is to assert that the `since` is not beyond the upper. If it already is, someone, somewhere must have messed up in driving around their `SinceHandle`.

Unless @petrosagg and you already discussed this and have more clever takes than I currently have. :sweat_smile:

Reply:
This ensures that the `upper` is also in advance of the remap collection's `SinceHandle`; it seems odd to assert that this must be true given that we never expressed a direct relationship between them previously. Can I get a little more detail on why/where that invariant is held?

Reply:
The requirement is not so much that the `upper` (of the data collection) needs to be in advance of the `since` of the remap shard. It's that the `since` of the remap shard must not go up to (or beyond) the `upper` of the data shard.

I know they sound similar, but one holds the `upper` fixed and varies the other, and vice versa.

It's critical that the `since` of the remap collection never reaches the `upper` of the data shard(s), because we otherwise cannot re-start ingestion.

Before your PR, the since of the remap collection is controlled by this: https://github.com/MaterializeInc/materialize/blob/e9e35a93b6cf4ba2d293355c1a73265103d6e9f6/src/storage/src/source/source_reader_pipeline.rs#L842

We assert on this invariant here: https://github.com/MaterializeInc/materialize/blob/917ce78e739e6f8b8b41e602f33e6a8a6646d805/src/storage/src/source/reclock/compat.rs#L107

The `as_of` that we see in that latter piece of code is the restart upper/frontier as calculated by the one implementation of `ResumptionFrontierCalculator`, here:
https://github.com/MaterializeInc/materialize/blob/af0b1406090c05961128ffe02dd29f8b1876f3fc/src/storage-client/src/types/sources.rs#L153

@petrosagg to double-check, please :pray:

Reply:
Ah, ok! When we had chatted about moving compaction from the reclock/remap operators into the storage controller, I had not completely followed the interaction with the resumption frontier. I'm comfortable closing this PR, but will wait for Petros to double-check the explanation here.